File: .agents/skills/yss-ddd-scaffold-generator/scripts/run_scaffold_verification.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""在生成项目根目录实际执行 YSS 脚手架的固定验证命令并留证。"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List


COMMANDS = ("validate", "test", "package")


def iso_now() -> str:
    return datetime.now(timezone.utc).isoformat()


def write_text(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8", errors="replace")


def run(project_root: Path, evidence_dir: Path) -> Dict[str, object]:
    wrapper = project_root / "mvnw"
    if not wrapper.is_file():
        raise FileNotFoundError(f"项目根目录缺少 Maven wrapper: {wrapper}")
    manifest_path = project_root / ".yss" / "scaffold-generation.json"
    if not manifest_path.is_file():
        raise FileNotFoundError(f"项目根目录缺少脚手架生成元数据清单: {manifest_path}")
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    required_manifest_fields = [
        "schema_version",
        "contract_id",
        "contract_version",
        "slice_id",
        "approval_ref",
        "approver",
        "lifecycle_approval_ref",
        "router_draft_ref",
        "persisted_ref",
        "contract_file_ref",
        "current_version",
        "allowed_write_paths",
        "expected_evidence_files",
        "verification_commands",
        "generation_mode",
    ]
    missing_manifest_fields = [
        field for field in required_manifest_fields if manifest.get(field) in (None, "")
    ]
    if manifest.get("generation_mode") != "controlled-generation" or missing_manifest_fields:
        raise ValueError(
            "脚手架生成元数据清单不完整或不是 controlled-generation: "
            + ", ".join(missing_manifest_fields)
        )
    if manifest["current_version"] != manifest["contract_version"]:
        raise ValueError("脚手架生成元数据清单不是当前合同版本")
    if manifest["verification_commands"] != [
        "./mvnw validate",
        "./mvnw test",
        "./mvnw package",
    ]:
        raise ValueError("脚手架生成元数据清单验证命令不符合固定合同")

    evidence_dir.mkdir(parents=True, exist_ok=True)
    command_results: List[Dict[str, object]] = []
    for phase in COMMANDS:
        command = f"./mvnw {phase}"
        stdout_path = evidence_dir / f"mvnw-{phase}.stdout.log"
        stderr_path = evidence_dir / f"mvnw-{phase}.stderr.log"
        started_at = iso_now()
        started = time.monotonic()
        try:
            completed = subprocess.run(
                [str(wrapper), phase],
                cwd=project_root,
                text=True,
                capture_output=True,
                check=False,
            )
            exit_code = completed.returncode
            stdout = completed.stdout
            stderr = completed.stderr
        except OSError as exception:
            exit_code = 127
            stdout = ""
            stderr = str(exception)
        duration_ms = round((time.monotonic() - started) * 1000, 3)
        write_text(stdout_path, stdout)
        write_text(stderr_path, stderr)
        command_results.append(
            {
                "command": command,
                "phase": phase,
                "exit_code": exit_code,
                "duration_ms": duration_ms,
                "started_at": started_at,
                "executed_at": iso_now(),
                "stdout_ref": str(stdout_path),
                "stderr_ref": str(stderr_path),
            }
        )

    return {
        "verification_mode": "controlled-generation",
        "project_root": str(project_root),
        "scaffold_manifest_ref": str(manifest_path),
        "generated_at": iso_now(),
        "status": "passed" if all(item["exit_code"] == 0 for item in command_results) else "failed",
        "commands": command_results,
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="运行 YSS 脚手架真实验证命令")
    parser.add_argument("--project-root", required=True, help="已生成后端项目的根目录")
    parser.add_argument("--evidence-dir", required=True, help="验证日志和 JSON 报告目录")
    args = parser.parse_args()

    project_root = Path(args.project_root).resolve()
    evidence_dir = Path(args.evidence_dir).resolve()
    report_path = evidence_dir / "scaffold-verification.json"
    try:
        report = run(project_root, evidence_dir)
    except Exception as exception:
        report = {
            "verification_mode": "controlled-generation",
            "project_root": str(project_root),
            "generated_at": iso_now(),
            "status": "failed",
            "error": str(exception),
            "commands": [],
        }
        report_path.parent.mkdir(parents=True, exist_ok=True)
        report_path.write_text(
            json.dumps(report, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
        )
        print(f"❌ 脚手架验证无法执行: {exception}", file=sys.stderr)
        return 1

    report_path.write_text(
        json.dumps(report, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )
    print(f"{report['status']}: {report_path}")
    return 0 if report["status"] == "passed" else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: .agents/skills/yss-frontend-scaffold-generator/scripts/run_scaffold_verification.mjs
#!/usr/bin/env node
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { runCommandSync } from "../../../../scripts/lib/command-runner.mjs";

const ALLOWED = new Map([
  ["pnpm install --frozen-lockfile", ["install", "--frozen-lockfile"]],
  ["pnpm lint", ["lint"]],
  ["pnpm type-check", ["type-check"]],
  ["pnpm build", ["build"]],
]);
function argsOf(argv) { const args = {}; for (let i = 0; i < argv.length; i += 2) args[argv[i].replace(/^--/, "")] = argv[i + 1]; return args; }
export function verify(projectRoot, evidenceDir, { timeoutMs = 0 } = {}) {
  projectRoot = path.resolve(projectRoot); evidenceDir = path.resolve(evidenceDir);
  const manifestPath = path.join(projectRoot, ".yss", "scaffold-generation.json");
  const manifest = JSON.parse(readFileSync(manifestPath, "utf8"));
  if (manifest.schema_version !== 4 || manifest.kind !== "frontend-scaffold" || manifest.generation_mode !== "controlled-generation") throw new TypeError("invalid frontend scaffold manifest");
  if (!manifest.generated_files.includes("pnpm-lock.yaml")) throw new TypeError("frontend template must contain pnpm-lock.yaml");
  const commands = manifest.verification_commands;
  for (const required of ALLOWED.keys()) if (!commands.includes(required)) throw new TypeError(`missing required verification command: ${required}`);
  mkdirSync(evidenceDir, { recursive: true });
  const results = commands.map((command, index) => {
    const argv = ALLOWED.get(command);
    if (!argv) throw new TypeError(`unsupported verification command: ${command}`);
    const started = Date.now();
    const stdoutRef = path.join(evidenceDir, `${index + 1}-stdout.log`), stderrRef = path.join(evidenceDir, `${index + 1}-stderr.log`);
    const result = runCommandSync("pnpm", argv, { cwd: projectRoot, timeoutMs, stdoutFile: stdoutRef, stderrFile: stderrRef, progress: true });
    return { command, exit_code: result.status ?? 1, duration_ms: Date.now() - started, ...(result.termination ? { termination: result.termination } : {}), stdout_ref: stdoutRef, stderr_ref: stderrRef, executed_at: new Date().toISOString() };
  });
  const passed = results.every((item) => item.exit_code === 0);
  const report = { schema_version: 1, kind: "frontend-scaffold-verification", status: passed ? "passed" : "failed", project_root: projectRoot, scaffold_manifest_ref: manifestPath, commands: results };
  const reportPath = path.join(evidenceDir, "scaffold-verification.json");
  writeFileSync(reportPath, `${JSON.stringify(report, null, 2)}\n`);
  if (passed) { manifest.completion_level = "empty-scaffold-verified"; manifest.scaffold_verification_ref = reportPath; writeFileSync(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`); }
  return report;
}
if (import.meta.url === `file://${process.argv[1]}`) {
  try { const args = argsOf(process.argv.slice(2)); const report = verify(args["project-root"], args["evidence-dir"], { timeoutMs: Number(args["timeout-ms"] || 0) }); process.stdout.write(`${report.status}\n`); if (report.status !== "passed") process.exitCode = 1; }
  catch (error) { process.stderr.write(`${error.message}\n`); process.exitCode = 1; }
}
